### PR TITLE
Add menu_image_icon check to detect raster image menu icons

### LINKED ISSUE #788: Add a check for images instead of icons in the main menu
_I know this is a chicken-and-egg-problem._

2020 I asked for a rule to disallow images in the main menu:
https://github.com/WordPress/wporg-plugin-guidelines/issues/71

It was closed because it wouldn't be enforceable.

Maybe adding this to the plugin check plugin could be a start to at least _recommend_ not using images, as they are often conflicting with one of the possible color schemes.

Best practice would be to use a dashicon or a custom SVG that will use the color from the color scheme used.

If you only want to add checks stated in the guidelines, but the guidelines don't want to add them, because they are not enforceable, we will have no progress here.

What do you think?

### PR DESCRIPTION
**Pull request by faisalahammad**
## What?
Closes #788

Adds a new warning-level check, menu_image_icon, that detects when plugins use raster image files (PNG, JPG, GIF, WebP, ICO, BMP) as the icon parameter in add_menu_page(). Raster images do not adapt to the WordPress admin color schemes, so a dashicon or an SVG data: URI is recommended instead.

## Why?
Raster image icons in the admin menu do not recolor to match the active admin color scheme, unlike dashicons and SVG data: URIs. This produces inconsistent admin UI. The check warns plugin authors about this best practice so they can switch to a color-scheme-safe icon.

## How?
The check extends Abstract_File_Check and uses a PHP token-based parser to locate add_menu_page() calls. It tracks open and close parens to reliably extract the sixth parameter, the icon URL, and reports a warning when the value is a raster image file. Dashicon classes, SVG data: URIs, empty strings and the 'none' value are all valid and skipped. The check is registered in Default_Check_Repository under the plugin_repo category.

## Testing Instructions
1. Install and activate Plugin Check.
2. Run Plugin Check on a plugin whose add_menu_page() call uses an image file as the icon, for example 'img/icon.png'.
3. Confirm a warning appears on the add_menu_page() call: "Raster image used as admin menu icon."
4. Run Plugin Check on a plugin that uses a dashicon or an SVG data: URI as the icon.
5. Confirm no warning is raised for the valid icons.

## AI Usage Disclosure

- [ ] This PR was created without the help of AI tools
- [x] This PR includes AI-assisted code or content

If AI tools were used, please describe how they were used:
Used Claude Code to implement the check, write the unit tests, and prepare this PR.

## Screenshots or screencast 

Not applicable. This change does not alter any user-facing UI.

<a href="https://playground.wordpress.net?blueprint-url=data:application/json,%7B%22%24schema%22%3A%22https%3A%2F%2Fplayground.wordpress.net%2Fblueprint-schema.json%22%2C%22landingPage%22%3A%22%2Fwp-admin%2Ftools.php%3Fpage%3Dplugin-check%22%2C%22phpExtensionBundles%22%3A%5B%22kitchen-sink%22%5D%2C%22steps%22%3A%5B%7B%22step%22%3A%22login%22%2C%22username%22%3A%22admin%22%2C%22password%22%3A%22password%22%7D%2C%7B%22step%22%3A%22installPlugin%22%2C%22pluginZipFile%22%3A%7B%22resource%22%3A%22url%22%2C%22url%22%3A%22https%3A%2F%2Fgithub.com%2FWordPress%2Fplugin-check%2Freleases%2Fdownload%2Fci-artifacts%2Fpr-1438-e052839ab352d7b524daa6aaadd90355904676b5.zip%22%7D%2C%22options%22%3A%7B%22activate%22%3Atrue%7D%7D%5D%7D" target="_blank" rel="noopener noreferrer">
  Open WordPress Playground Preview
</a>

### CODE REVIEW

**Comment by faisalahammad** [comment]
## CI Fix Summary - 1 failure resolved

| # | File | Error | Fix |
|---|------|-------|-----|
| 1 | Menu_Image_Icon_Check.php:124 | PHPMD NPathComplexity 2841>200 + CyclomaticComplexity 20>=20 | Added @SuppressWarnings(PHPMD.CyclomaticComplexity) + @SuppressWarnings(PHPMD.NPathComplexity), matching existing repo pattern |

Tests ok · Verification ok · PHP 8.5 ok

File: includes/Checker/Checks/Plugin_Repo/Menu_Image_Icon_Check.php [hunk 1]
```diff
@@ -0,0 +1,304 @@
...
+			);
+		}
+	}
+
+	/**
+	 * Scans PHP files for add_menu_page() calls with a quoted string icon parameter.
+	 *
+	 * The icon URL is the sixth parameter of add_menu_page(). A regex is used to count
+	 * to the sixth parameter, capturing the icon string along with the file position.
+	 *
```
[the remark this answers is not in the document]
Comment (faisalahammad) [reply includes/Checker/Checks/Plugin_Repo/Menu_Image_Icon_Check.php:120]: Fixed in 52aa184. Updated docblock to accurately describe token-based parser instead of regex. Also fixed `file_contents()` docblock at line 233.

File: includes/Checker/Checks/Plugin_Repo/Menu_Image_Icon_Check.php [hunk 2]
```diff
@@ -0,0 +1,304 @@
...
+				$offset   += strlen( $text );
+			}
+
+			$count = count( $scanned );
+			for ( $index = 0; $index < $count; $index++ ) {
+				if ( T_STRING !== $scanned[ $index ]['id'] || 'add_menu_page' !== strtolower( $scanned[ $index ]['text'] ) ) {
+					continue;
+				}
+
+				$open = $index + 1;
```
[the remark this answers is not in the document]
Comment (faisalahammad) [reply includes/Checker/Checks/Plugin_Repo/Menu_Image_Icon_Check.php:170]: Fixed in 52aa184. Added backward scan for previous significant token: skips `$obj->add_menu_page()`, `Plugin_Manager::add_menu_page()`, and PHP 8+ `$obj?->add_menu_page()` calls. Test fixtures added for method and static call cases.

**Comment by davidperezgar** [comment]
Thanks to Codex’s review, I found one edge case:

On PHP 8+, `\add_menu_page()` is tokenized as `T_NAME_FULLY_QUALIFIED` rather than `T_STRING`, so explicitly global calls—commonly used in namespaced plugins—are currently skipped by this check.

Could we support that token type while keeping PHP 7.4 compatibility (for example, by guarding it with `defined()`), normalize the leading backslash, and add a regression test?

**Comment by faisalahammad** [comment]
Addressed @davidperezgar review: fully-qualified \add_menu_page() calls are now detected. On PHP 8+ they tokenize as T_NAME_FULLY_QUALIFIED; the check matches that (guarded with defined()/constant() for PHP 7.4) and normalizes the leading backslash. New namespaced regression fixture plus a fully-qualified case in the with-errors fixture.

PHPUnit passes on the PHP 7.4/8.0-8.4 matrix and phpcs/phpstan are clean. Typos failed only on a transient 503 binary download, not a spell finding. Ready for re-review (commit 64db1d1).

**Review by ernilambar — APPROVED** [review]

**Review by davidperezgar — APPROVED** [review]
